src/main/java/com/tabnine/statusBar/StatusBarUpdater.kt

```diff
@@ -0,0 +1,94 @@
+        ApplicationManager.getApplication().executeOnPooledThread { requestStatusBarMessage() }
+    }
+
+    fun requestStatusBarMessage() {
+        try {
+            if (!synchronizer.compareAndSet(false, true)) {
+                return // abort if there's already a request in progress
+            }
+            val statusBarPromotionWidgets = getStatusBarsWidgets()
+            if (statusBarPromotionWidgets.isEmpty()) {
+                return
+            }
+
+            val promotion = binaryRequestFacade.executeRequest(StatusBarPromotionBinaryRequest())
+
+            if (promotion != null) {
+                updateStatusBars(statusBarPromotionWidgets, promotion)
+                binaryRequestFacade.executeRequest(
+                    StatusBarPromotionShownRequest(
+                        promotion.id,
+                        promotion.message ?: "undefined", promotion.notificationType,
+                        promotion.state
+                    )
+                )
+            } else {
+                clear(statusBarPromotionWidgets)
+            }
+        } finally {
+            synchronizer.set(false)
+        }
+    }
+
+    private fun updateStatusBars(
+        statusBarPromotionWidgets: List<StatusBarPromotionWidget.StatusBarPromotionComponent>,
+        statusBarPromotionResponse: StatusBarPromotionBinaryResponse
+    ) {
+        updateStatusBars(
+            statusBarPromotionWidgets, true,
+            statusBarPromotionResponse.message,
+            statusBarPromotionResponse.id, statusBarPromotionResponse.actions,
+            statusBarPromotionResponse.notificationType
+        )
+    }
+
+    private fun clear(statusBarPromotionWidgets: List<StatusBarPromotionWidget.StatusBarPromotionComponent>) {
+        updateStatusBars(
+            statusBarPromotionWidgets, false, null,
+            null, null, null
+        )
+    }
+
+    private fun updateStatusBars(
+        statusBars: List<StatusBarPromotionWidget.StatusBarPromotionComponent>,
+        isVisible: Boolean,
+        text: String?,
+        id: String?,
+        actions: List<String>?,
+        notificationType: String?,
+    ) {
+        for (statusBarPromotionWidget in statusBars) {
+            statusBarPromotionWidget.isVisible = isVisible
+            statusBarPromotionWidget.text = text
+            statusBarPromotionWidget.id = id
+            statusBarPromotionWidget.actions = actions
+            statusBarPromotionWidget.notificationType = notificationType
+        }
+    }
+
+    private fun getStatusBarsWidgets(): List<StatusBarPromotionWidget.StatusBarPromotionComponent> {
+        return ProjectManager.getInstance().openProjects.mapNotNull {
+            val statusBar = WindowManager.getInstance()?.getStatusBar(it)
+            val widget = statusBar?.getWidget(StatusBarPromotionWidget::class.java.name) ?: null
+            (widget as StatusBarPromotionWidget).component as StatusBarPromotionWidget.StatusBarPromotionComponent
```

Comment:
Isn't that a possible bug? `widget` could be `null` and when you ask for `.component` it will through `NullPointerException`. Or am I missing something here?

Reply:
By the way, Kotlin does not yell on you here because you convert the widget to non nullable. If you'd do `(widget as StatusBarPromotionWidget?)` I think it would not compile

Reply:
You are correct. Wasn't sure about the correct syntax for this. Thanks.